hwmv2: intel_adsp: Port to HWMv2

golowanow opened the pull request:
Port `intel_adsp` SoC family: `ace` and `cavs` SoC series, and related board configurations:

* `intel_adsp_cavs25`
* `intel_adsp_cavs25_tgph` changed to `intel_adsp_cavs25/intel_tgl_adsp/tgph`
* `intel_adsp_ace15_mtpm`
* `intel_adsp_ace20_lnl`

cc:
- #51831

golowanow commented:
@tejlmand, @nashif, @dcpleung  - for your early feedback as a draft.

golowanow committed:
soc: xtensa: intel_adsp: move and convert to HWMv2

Move and convert soc/xtensa/intel_adsp SoC family configurations
to HWMv2 with its SoC series:
`ace` (INTEL_ACE) and `cavs` (INTEL_ADSP_CAVS).

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
boards: xtensa: intel_adsp_cavs25: move and convert to HWMv2

Move and convert to HWMv2 intel_adsp_cavs25 board configuration.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
boards: xtensa: intel_adsp_cavs25_tgph: change to board variant

Change `intel_adsp_cavs25_tgph` board definition to be HWMv2 board
variant `intel_adsp_cavs25/intel_tgl_adsp/tgph`

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
boards: xtensa: intel_adsp_ace15_mtpm: move and convert to HWMv2

Move and convert to HWMv2 intel_adsp_ace15_mptm board configuration.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
boards: xtensa: intel_adsp_ace20_lnl: move and convert to HWMv2

Move and convert to HWMv2 intel_adsp_ace20_lnl board configuration.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
1. adjusted to #68346 
2. `intel_adsp_cavs25_tgph` board configuration is converted to a HWMv2 revision `intel_adsp_cavs25/intel_tgl_adsp/tgph`

golowanow commented on soc/intel/intel_adsp/cavs/_soc_inthandlers.h:
this move from `intel_adsp/common/include` to `intel_adsp/cavs/` resolves build issue when this header file was applied with the HWMv2 for `adsp` builds as well

golowanow commented on boards/intel/intel_adsp_cavs25/board.yml:
do we really need to change `tgph` to `intel_adsp_cavs25` board variant ?

nashif commented:
> * `intel_adsp_cavs25_tgph` changed to `intel_adsp_cavs25/intel_tgl_adsp/tgph`

oh wow, not nice at all, @tejlmand, we need away for this to be shorter and to the point, lots of redundancy in that name.

nordicjm commented:
> > * `intel_adsp_cavs25_tgph` changed to `intel_adsp_cavs25/intel_tgl_adsp/tgph`
> 
> oh wow, not nice at all, @tejlmand, we need away for this to be shorter and to the point, lots of redundancy in that name.

As said before, if there is only one soc then you can omit the soc, i.e. `intel_adsp_cavs25//tgph`, that should only be omitted by users on the command line though, .yaml files and sample.yml/testcase.yml should use the full name

nordicjm commented on soc/intel/intel_adsp/Kconfig:
```diff
@@ -1,22 +1,17 @@
-	default "intel_adsp"
-
 # Select SoC Part No. and configuration options
-source "soc/soc_legacy/xtensa/intel_adsp/*/Kconfig.soc"
+# rsource "*/Kconfig.soc"
```
```suggestion
rsource "*/Kconfig"
```

golowanow replied:
changed

nordicjm commented on soc/intel/intel_adsp/ace/Kconfig.defconfig.series:
```diff
@@ -68,6 +60,6 @@ config LOG_BACKEND_ADSP
 endif # LOG
 
-source "soc/soc_legacy/xtensa/intel_adsp/ace/Kconfig.defconfig.ace*"
+# rsource "Kconfig.defconfig.ace*"
```
??

golowanow replied:
my bad, left while debugging

nordicjm commented on soc/intel/intel_adsp/ace/Kconfig.series:
```diff
@@ -14,3 +15,12 @@ config SOC_SERIES_INTEL_ACE
+	default "ace" if SOC_SERIES_INTEL_ACE
+
+config SOC_TOOLCHAIN_NAME
+	string
```
```suggestion
```

nordicjm commented on soc/intel/intel_adsp/ace/Kconfig.series:
```diff
@@ -1,8 +1,9 @@
-# Copyright (c) 2022 Intel Corporation
+# Copyright (c) 2022-2024 Intel Corporation
```
This should be Kconfig.soc

nordicjm commented on soc/intel/intel_adsp/ace/Kconfig.series:
```diff
@@ -1,8 +1,9 @@
-	bool "Intel ACE"
+	bool
 	select SOC_FAMILY_INTEL_ADSP
 	select XTENSA
```
These lines should be in the Kconfig file

golowanow replied:
changed

nordicjm commented on soc/intel/intel_adsp/ace/Kconfig.soc:
```diff
@@ -0,0 +1,26 @@
+
+config SOC_INTEL_ACE15_MTPM
+	bool
+	help
+	  ACE 1.5 Meteor Lake PCH M
```
help goes at bottom

golowanow replied:
done

nordicjm commented on soc/intel/intel_adsp/ace/Kconfig.soc:
```diff
@@ -0,0 +1,26 @@
+
+config SOC_INTEL_ACE20_LNL
+	bool
+	help
```
etc.

golowanow replied:
done

nordicjm commented on soc/intel/intel_adsp/cavs/Kconfig.series:
```diff
@@ -1,8 +1,9 @@
-# Copyright (c) 2017,2022 Intel Corporation
+# Copyright (c) 2017,2022-2024 Intel Corporation
```
All `Kconfig.series` files should be renamed to `Kconfig.soc`

golowanow replied:
changed

nordicjm commented on boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config  BOARD_INTEL_ADSP_CAVS25
```
double space after Kconfig

golowanow replied:
done

nordicjm commented on boards/intel/intel_adsp_cavs25/Kconfig.intel_adsp_cavs25_tgph:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config  BOARD_INTEL_ADSP_CAVS25_TGPH
```
etc

golowanow replied:
done

golowanow committed:
soc: xtensa: intel_adsp: HWMv2 workaround for SOF config

Apply a workaround at Kconfig.defconfig for CORE_COUNT default
value set overriding SOF arch/host incorrect defaults chosen
without ARCH specified by the changed HWMv2 cmake sequence.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

nashif commented:
> As said before, if there is only one soc then you can omit the soc, i.e. `intel_adsp_cavs25//tgph`, that should only be omitted by users on the command line though, .yaml files and sample.yml/testcase.yml should use the full name

as said before, this context aware board naming and guessing is not ideal and is going to be very error-prone. The fact is that for reference we will need one unique and authorative name for boards and should not rely on the context, i.e. for tooling and what not. Such long names with lots of redundancy are not user or developer friendly at all. I am sure this can be solved in the definition of the board or soc itself, but we are giving a people a rope to hang themselves with and we will end up with 1m long board names with lots of redunancy that is not as intuitive as it may seem.

nashif commented:
@tejlmand  @nordicjm  btw, there is something very wrong here, SOF failing to build and picking kconfigs from some completely unrelated _defconfig file (for CONFIG_MULTICORE defined in SOF). Looks like some handling of ARCH is not quite right.

golowanow commented:
> @tejlmand @nordicjm btw, there is something very wrong here, SOF failing to build and picking kconfigs from some completely unrelated _defconfig file (for CONFIG_MULTICORE defined in SOF). Looks like some handling of ARCH is not quite right.

the proposed fix will be in this PR soon

golowanow commented on soc/intel/intel_adsp/Kconfig.defconfig:
where CORE_COUNT is set to default 1 at SOF module when its Kconfig is processed by the HWMv2 altered build sequence (without ARCH set):
https://github.com/zephyrproject-rtos/sof/blob/0606152d4aafc1f7ed43df1b1813252bfc74e154/src/arch/host/Kconfig#L5-L9

so the workaround's idea is to just re-align to MAX_CORE_COUNT of the SoC similar to
https://github.com/zephyrproject-rtos/sof/blob/0606152d4aafc1f7ed43df1b1813252bfc74e154/src/platform/Kconfig#L240-L247

golowanow commented:
@tejlmand, @nashif, @dcpleung - up to your attention.

golowanow commented:
adjusted to the `collab-hwm` rebase

golowanow commented:
fix `KconfigHWMv2` check issues of this PR, but still needs #68811 to run it on CI

golowanow committed:
scripts: west: runners: intel_adsp: Adjust path to HWMv2

Adjust script path to HWMv2 location at `soc/intel/intel_adsp/tools`

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

nashif commented:
Thought about it a bit and played with this PR, I think it makes much more sense to have a revised structure that will give us something like this:

```

  Valid board identifiers for intel_adsp are:

  intel_adsp/cavs25

  intel_adsp/cavs25/tgph

  intel_adsp/ace15_mtpm

  intel_adsp/ace20_lnl
```

So basically have the board definition go under intel_adsp and revise soc name to match v1 naming, this will give us a very close naming to what we had before and will make it future proof.

golowanow commented:
rebase and few more path changes in cavs25 documentation and west runner script.

golowanow committed:
soc: boards: xtensa: intel_adsp_cavs25: Rename with HWMv2

Rename for better HWMv2 use:
 SoC - `intel_tgl_adsp` to `cavs25`

Resulting name changes for boards:
 `intel_adsp_cavs25/intel_tgl_adsp`      -> `intel_adsp_cavs25/cavs25`
 `intel_adsp_cavs25/intel_tgl_adsp/tgph` -> `intel_adsp_cavs25/cavs25/tgph`

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
boards: xtensa: intel_adsp_cavs25: Rename board with HWMv2

Rename for better HWMv2 use:
 board - `intel_adsp_cavs25` to `intel_adsp`

Resulting name changes for boards:
 `intel_adsp_cavs25/cavs25`      -> `intel_adsp/cavs25`
 `intel_adsp_cavs25/cavs25/tgph` -> `intel_adsp/cavs25/tgph`

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
tests: samples: boards: intel_adsp_cavs25: Rename with HWMv2

Adjust tests and samples to HWMv2 board rename:
     `intel_adsp_cavs25`      -> `intel_adsp/cavs25`
     `intel_adsp_cavs25_tgph` -> `intel_adsp/cavs25/tgph`

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
doc: soc: boards: intel_adsp_cavs25: Rename with HWMv2

Adjust documentation to HWMv2 board rename:
  `intel_adsp_cavs25`      -> `intel_adsp/cavs25`
  `intel_adsp_cavs25_tgph` -> `intel_adsp/cavs25/tgph`

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
> I think it makes much more sense to have a revised structure that will give us something like this:
> 
> ```
>   intel_adsp/cavs25
>   intel_adsp/cavs25/tgph
> ```
> 
> So basically have the board definition go under intel_adsp and revise soc name to match v1 naming, this will give us a very close naming to what we had before and will make it future proof.

@nashif  I've added few commits to rename `cavs25` is it what you think we should do for it,  same for `ace` ?
@nordicjm can you also look ?

nashif commented:
> and now it can't run on CI here because of platform name changed:
> 
> ```
> ERROR   - sample.audio.sof - integration_platforms - unrecognized platform - intel_adsp_cavs25
> ```

that is expected, we need to change the yaml file in sof, so submit a corrosponding PR to the sof project and link it in here in optional manifest using the PR to https://github.com/zephyrproject-rtos/sof

golowanow committed:
hwmv2: sof: intel_adsp: submanifest provisional link

Provisional link to SOF with adjustment to the HWMv2 new
intel_adsp board names.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

nashif commented on submanifests/optional.yaml:
```diff
@@ -34,7 +34,7 @@ manifest:
         - optional
     - name: sof
-      revision: 0606152d4aafc1f7ed43df1b1813252bfc74e154
+      revision: b9d637c1b31e597064b7b946f4f18fb73e3dca5b
```
you should be pointing to a PR, not a commit sha

golowanow commented:
> > I think it makes much more sense to have a revised structure that will give us something like this:
> > ```
> >   intel_adsp/cavs25
> >   intel_adsp/cavs25/tgph
> > ```
> >     
> >   
> > So basically have the board definition go under intel_adsp and revise soc name to match v1 naming, this will give us a very close naming to what we had before and will make it future proof.
> 
> @nashif I've added few commits to rename `cavs25` is it what you think we should do for it, same for `ace` ? @nordicjm can you also look at this ?

@nashif so moving all boards to this scheme or keep v1 for now  ?

golowanow committed:
MAINTAINERS: intel_adsp: Adjust to HWMv2 move

Adjust intel_adsp paths to HWMv2 move.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
soc: boards: xtensa: intel_adsp_ace: Rename with HWMv2

Rename for better HWMv2 use:

  SoC - `intel_ace15_mtpm` --> `ace15_mtpm`
  SoC - `intel_ace20_lnl`  --> `ace20_lnl`

Resulting shortened name for boards:

  `intel_adsp_ace15_mtpm/intel_ace15_mtpm` -->
  `intel_adsp_ace15_mtpm/ace15_mtpm`

  `intel_adsp_ace20_lnl/intel_ace20_lnl` -->
  `intel_adsp_ace20_lnl/ace20_lnl`

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
hwmv2: boards: intel_adsp: Overhaul board configurations

Rename and join configurations for intel_adsp CAVS and ACE
boards to benefit from HWMv2 shortened names with resulting
names change:

  `intel_adsp_ace15_mtpm` --> `intel_adsp/ace15_mtpm`
  `intel_adsp_ace20_lnl`  --> `intel_adsp/ace20_lnl`

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
hwmv2: tests: boards: intel_adsp: Adjust board names

Adjust tests to HWMv2 intel_adsp_ace board name changes.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
> > >
```
  intel_adsp/cavs25
  intel_adsp/cavs25/tgph
  intel_adsp/ace15_mtpm
  intel_adsp/ace20_lnl
``` 
> > > So basically have the board definition go under intel_adsp and revise soc name to match v1 naming, this will give us a very close naming to what we had before and will make it future proof.
> > 
> 
> @nashif so moving all boards to this scheme or keep v1 for now ?

all `intel_adsp` boards are renamed to the proposed shortened form.

golowanow committed:
soc: intel_adsp: tools: pylint compliance workaround

Disable Pylint compliance check warning `R0801:Similar lines`.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow committed:
doc: boards: intel_adsp: Re-order pages

Index boards/intel/intel_adsp pages.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented:
re-assign reviewers according to the current MANIFEST

tejlmand commented on boards/intel/intel_adsp/Kconfig.intel_adsp:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_INTEL_ADSP_CAVS25
+	bool
+	select SOC_INTEL_CAVS_V25
+
+config BOARD_INTEL_ADSP_CAVS25_TGPH
+	bool
+	select SOC_INTEL_CAVS_V25
+
+config BOARD_INTEL_ADSP_ACE15_MTPM
+	bool
+	select SOC_INTEL_ACE15_MTPM
+
+config BOARD_INTEL_ADSP_ACE20_LNL
+	bool
+	select SOC_INTEL_ACE20_LNL
```
Please, no bool, as the `BOARD_<name>` and `BOARD_<full_identifier>` are constructed here:
https://github.com/zephyrproject-rtos/zephyr/blob/db742d03aa8fa469a96435fe85b292e62d061d38/boards/Kconfig.v2#L8
https://github.com/zephyrproject-rtos/zephyr/blob/db742d03aa8fa469a96435fe85b292e62d061d38/boards/Kconfig.v2#L13

Some details can be found here: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#common-kconfig-shorthands

Please do:
```suggestion
config BOARD_INTEL_ADSP
	select SOC_INTEL_CAVS_V25 if BOARD_INTEL_ADSP_CAVS25
	select SOC_INTEL_CAVS_V25 if BOARD_INTEL_ADSP_CAVS25_TGPH
	select SOC_INTEL_ACE15_MTPM if BOARD_INTEL_ADSP_ACE15_MTPM
	select SOC_INTEL_ACE20_LNL if  BOARD_INTEL_ADSP_ACE20_LNL
```
as that will ensure any typos or similar wrong selection is caught by CI.

and it's fewer lines 😉

golowanow replied:
> Please, no bool, as the `BOARD_<name>` and `BOARD_<full_identifier>` are constructed here:

well, as I recall, without these `bool` it fails on build as config variables has no type for other boards except the current one.
@tejlmand m.b. there is some other problem with this way of board definition change then ?

golowanow replied:
> and it's fewer lines 😉

right, that's it, thank you, changed.

tejlmand reviewed (changes requested):
Thanks for this porting, but please update the board's Kconfig.

golowanow commented on boards/intel/intel_adsp/board.cmake:
```diff
@@ -0,0 +1,42 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_BOARD_INTEL_ADSP_CAVS25 OR CONFIG_BOARD_INTEL_ADSP_CAVS25_TGPH)
```
it has been re-composed from 3 `board.cmake` files, so it is rather created anew.

tejlmand replied:
> it has been re-composed from 3 `board.cmake` files, so it is rather created anew.

in that case fixing indentation would be nice.

One more case of `git` not really supporting renames, but instead relies on the similarity index matching 😉 
```
diff --git a/boards/intel/intel_adsp_cavs25/board.cmake b/boards/intel/intel_adsp/board.cmake
similarity index 90%
rename from boards/intel/intel_adsp_cavs25/board.cmake
rename to boards/intel/intel_adsp/board.cmake
index 46692b0b028..1bdb2698c12 100644
--- a/boards/intel/intel_adsp_cavs25/board.cmake
+++ b/boards/intel/intel_adsp/board.cmake
...
```

tejlmand reviewed (approved):
approved, but as it turns out `board.cmake` is created new, then fixing indentation in the process would be nice, but can also be done later.

golowanow commented:
> approved, but as it turns out `board.cmake` is created new, then fixing indentation in the process would be nice, but can also be done later.

indentation fixed

ceolin commented on soc/intel/intel_adsp/Kconfig.defconfig:
```diff
@@ -1,11 +1,18 @@
+config CORE_COUNT
+	int
+	default MP_MAX_NUM_CPUS
+endif
```
Can't we get rid of this now since we already have to do changes in SOF side ?
